docs/design/service_app_protocol.md

```diff
@@ -0,0 +1,84 @@
+
+This document proposes to leverage the newly introduced `AppProtocol` field to determine the application protocol for a service’s port in OSM. The application protocol on the port will be used to infer the protocol of the inbound traffic received on that port by the upstream proxy. Based on the protocol detected, appropriate filer chains will be configured on the upstream proxy to apply protocol specific filters to handle protocol specific policies.
+
+The minimum Kubernetes version must be v1.19 for features that require the `AppProtocol` to be specified for the service ports. If the Kubernetes version is lower than v1.19 or if the `AppProtocol` field is not set on the service's ports, OSM controller will log error events where applicable.
+
+
+## Example
+
+Consider the following SMI traffic access and traffic specs policies:
+- A `TCPRoute` resource named `tcp-route` that specifies the port TCP traffic should be allowed on.
+- An `HTTPRouteGroup` resource named `http-route` that specifies the HTTP routes for which HTTP traffic should be allowed.
+- A `TrafficTarget` resource named `test` that allows pods in the service account `sa-2` to access pods in the service account `sa-1` for the speficied TCP and HTTP rules.
+
+```yaml
+kind: TCPRoute
+metadata:
+  name: tcp-route
+spec:
+  matches:
+    ports:
+    - 8080
+---
+kind: HTTPRouteGroup
+metadata:
+  name: http-route
+spec:
+  matches:
+  - name: version
+    pathRegex: "/version"
+    methods:
+    - GET
+---
+kind: TrafficTarget
+metadata:
+  name: test
+  namespace: default
+spec:
+  destination:
+    kind: ServiceAccount
+    name: sa-1
+    namespace: default
+  rules:
+  - kind: TCPRoute
+    name: tcp-route
+  - kind: HTTPRouteGroup
+    name: http-route
+  sources:
+  - kind: ServiceAccount
+    name: sa-2
+    namespace: default
+```
+
+Kubernetes service resources should explicitly specify the application protocol being served by the service's ports using the `appProtocol` field.
+
+A service `service-1` backed by a pod in service account `sa-1` serving `http` application traffic should be defined as follows:
+
+```yaml
+kind: Service
+metadata:
+  name: service-1
+  namespace: default
+spec:
+  ports:
+  - port: 8080
+    name: some-port
+    appProtocol: http
+```
+
+A service `service-2` backed by a pod in service account `sa-1` serving raw `tcp` application traffic shold be defined as follows:
+
+```yaml
+kind: Service
+metadata:
+  name: service-1
+  namespace: default
+spec:
+  ports:
+  - port: 8080
+    name: some-port
+    appProtocol: tcp
```

Comment:
```suggestion
    appProtocol: tcp # Same port number as service-1; different protocol
```

Reply:
This need not be true. I don't want to be introducing such assumptions as a part of this PR.